server-protocol.md

```diff
@@ -104,6 +114,45 @@ was sent in the wrong order), an `error` response will be sent, This response
 will include the error string in the `error` key, and a full copy of the
 original message dictionary in `orig`.
 
+
+## Permission and Abilities
+
+Server operators may wish to deny service to some clients.
+
+One such use-case is if the server is under a Denial of Service (DoS)
+attack or other malicious activity.
+
+In the `welcome` message the server may include a `permission-required` key. If provided, this will point to a `dict` with one key per supported method along with any method-specific options.
+For example::
+
+    {
+        "none": {}
+        "hashcash": {
+            "bits": 6,
+            "resource": "arbitrary string"
+        }
+    }
+
+Currently, the following are supported by the protocol:
+
+* `none`: no permission required, send a normal `bind`.
+* `hashcash`: Includes keys for `bits` and `resource` used as input as per the
```

Comment:
I'm not 100% sure how hashcash works. Because as far as I can tell, it works like "I create a token that proves that I did work". But this means that the server now needs to check that tokens aren't used multiple times. I'd prefer a challenge-response style method where the server issues a (unique and individual) challenge to solve. Then, there is no need to check for token re-use. But maybe hashcash supports this and I am misunderstanding something somewhere …

Reply:
I believe we can use the "resource string" for that (e.g. we could include a timestamp plus a nameplate for example). We do need to decide upon that though, but I think it can be made to work such that we _don't_ need a bloom filter or whatever to track "spent" tokens. We could even just include a random HMAC (and then the server can verify that _it_ produced the random string that's in the incoming hashcash).

Reply:
Apparently hashcash thought of this by using timestamps. In our case, we can indeed use the "resource string" for that purpose. (I'd suggest a simple counter on the server side.)

Reply:
There is an "ext" field in hashcash that is unused in the version 1 (current version). We could even make use of it to send back an HMAC or something to send a response to the challenge.

Reply:
>  We could even just include a random HMAC (and then the server can verify that _it_ produced the random string that's in the incoming hashcash).

That doesn't protect against re-use, though. When I worked through this for Petmail, I think I settled upon:
* server-provided `resource` is a parseable combination of timestamp and random string
* server remembers set of (timestamp+string) pairs
* when stamp is submitted, server rejects if (timestamp < cutoff) or the pair is in the remembered set
* periodically (maybe once per hour), server updates cutoff time, walks remembered set, deletes all entries with timestamp < cutoff
* in general, the cutoff is an hour or two older than the current timestamp, to allow the client some hashing/turnaround time
* if the set grows too large, the server should either increase the work factor (to reduce the rate of successful additions to the set), or simply deny all access until the periodic pruning has had a chanec to delete some entries, to prevent a RAM exhaustion attack via successful work
*